### DISCUSSION TITLE
Convenient script to install Jamulus on a Debian cloud server

### DISCUSSION THREAD

**Discussion by dtinth** [jamulussoftware/jamulus#1961] [General]
I recently met group of blind musicians who are jamming on our public server, and helped them set up their own private Jamulus server on the cloud, for their livestreaming sessions. Trying to follow the instructions on https://jamulus.io/wiki/Server-Linux proves to be challenging:

- The instructions are hard to follow without prior Linux knowledge, e.g. how to download the .deb file and put it on a server.
- Changing the server configuration requires using a text editor. This can be challenging, especially for a blind user to do on their own.

So inspired by convenient scripts like [openvpn-install](https://github.com/Nyr/openvpn-install), I created a script [`ez-jamulus-server`](https://github.com/dtinth/ez-jamulus-server/blob/main/ezjaminstall.sh). Here is how to use it:

1. Launch a **Debian 10** VM on a cloud provider of choice.
2. SSH into the machine and run `wget -O- https://git.io/ezjam | sudo bash` (many cloud providers now provide SSH over web browser)
3. Server is online and can be connected.

The [documentation](https://github.com/dtinth/ez-jamulus-server/blob/main/docs/modules/ROOT/pages/index.adoc) provides commands to change the server name, location, welcome message, and maximum number of users. Once the configuration is changed, re-running the installation script will restart the server.

I hope you find this useful. Please note that I currently have no intention to keep these scripts up-to-date, so if the community would like to maintain this script, please feel free to fork the project, and I will point people to that project.

**Comment by ann0see** [comment]
Great idea! I think @niebert, @jp8, @softins and some others already created server-installation scripts. Probably they should be unified. 

This repo https://github.com/jamulussoftware/installscripts is somewhat unmaintained and not in focus. Maybe you can update the scripts and raise awareness on them?

**Reply by dtinth** [reply]
Thank you for pointing out the installscripts repo, I am totally unaware of it!

I currently have no plans to maintain the scripts, but if anyone wants to copy mine, please feel free. If the installscripts becomes up-to-date and is easy enough to use (see below), I will point people to that script and update my script to redirect there.

In case it may be helpful, here are some considerations taken when I created the script:

1. The installation script should be short and easy enough to type by hand.

    <details><summary>Why?</summary>

    At first the installation script looks like this:

    ```
    wget -O- https://gist.githubusercontent.com/dtinth/b8d23de26a1820ec6f7a7238bc895f44/raw/jamulus-debian.sh | sudo bash
    ```

    However, there are some challenges for users to run this:

    - A screen reader user has trouble pasting text into a PuTTY session.
    - Some cloud provider’s web-based console access doesn’t allow pasting text; all commands must be typed in.

    </details>

2. A user should be able to change server configuration without requiring usage of text editor.

**Comment by gilgongo** [comment]
Thanks for doing this, which obviously helps not only blind users but those who don't know much about Linux as well.

BTW there have occasionally been calls to make the Linux docs more explanatory for those not familar with the OS. In my experience, that's probably a bad idea though. This is becuase the huge variety of Linux configurations and variations between distributions and hosting environments makes the work of accomodating even the more common situations counter-productive: attempts to cater for variations very rapidly increases complexity of the text so that maintaining - and more importantly reading - the docs becomes difficult and error prone. This then leads to more calls for explanation or simplification as more people are confused. A vicious cycle. 

The current docs are written mostly with Debian/Ubuntu in mind, and attempt to steer a "common denomiator" through that. Yes, this means the reader nees to be reasonably knowledgable, but then again, does running a server on the public Internet imply a level of competance to manage security and other issues outside of Jamulus setup?

**Reply by dtinth** [reply]
Thanks for the comment. IMO, for people who just want to get the job done, just providing a “for dummies” guide that covers 80% of the use case can be very useful already. I don’t think everyone who runs a server needs to know about Linux server administration. 

Here in Thailand, running a server on a personal machine may not be very practical, because the internet infrastructure in Thailand makes some server unreachable. Before coming up with this script, I helped several people setting up a server on their home machine. They did port forwarding and everything, but the server remains unreachable by a portion of the jammers.

So setting up a server on the cloud seems to be the only reliable way here to connect musicians together.

> does running a server on the public Internet imply a level of competance to manage security and other issues outside of Jamulus setup?

I would say not necessarily, since the server is created to run Jamulus only, and is destroyed after use... In other words, it’s disposable. Here’s how I guide people to set up a Jamulus server.

1. Create a cloud server on a provider (like Linode, or DigitalOcean). When asked to choose an OS, select Debian 10. Type in a root password. Disable all kinds of firewall.

2. Open a web-based console, log in, and run the appropriate commands on the server.

3. Once finished using, destroy the server, otherwise the provider will charge every hour.

**Reply by gilgongo** [reply]
> just providing a “for dummies” guide that covers 80% of the use case can be very useful already

There are a number of third party guides to setting up Jamulus servers on specific platforms ([such as this one](https://www.facebook.com/notes/jamulus-worldjam/howto-idiots-guide-to-installing-or-upgrading-a-jamulus-server-on-amazon-aws-lig/818091045662521/) for Amazon Lightsail).  

As to your specific outline of how to create and take down a VPS (reluctant as I am in taking the standard "FOSS project deflection" on this) - none of that relates to setting up and managing Jamulus itself. The question is therefore where you draw the line. 

Incidentally, there is/are third party systems for spinning up temporary Jamulus servers - no hosting or knowledge of anything other than the Jamulus client is needed (I can't remember the name of the main one now...). The down side may be lack of worldwide locations though.
